### PR TITLE
Spoke Accrue Interest Tests

### LINKED ISSUE #144: Spoke RP - Accrue Interest Unit Tests

### PR DESCRIPTION
**Pull request by CheyenneAtapour**
Closes #144

### CODE REVIEW

File: tests/Base.t.sol [hunk 1]
```diff
@@ -812,20 +812,15 @@ abstract contract Base is Test {
...
-    // asset.suppliedShares = hub.getAssetSuppliedShares(assetId);
-    // asset.availableLiquidity = hub.getAvailableLiquidity(assetId);
-    // (asset.baseDebt, asset.outstandingPremium) = hub.getAssetDebt(assetId);
-    // asset.baseBorrowIndex = hub.getAsset(assetId).baseBorrowIndex;
-    // asset.baseBorrowRate = hub.getBaseInterestRate(assetId);
-    // asset.riskPremium = hub.getAssetRiskPremium(assetId);
-    // asset.lastUpdateTimestamp = hub.getAsset(assetId).lastUpdateTimestamp;
-    // asset.config = hub.getAssetConfig(assetId);
-    // return asset;
+  function getSpokeInfo(
```
Comment (yan-man) [inline tests/Base.t.sol:873]: i was thinking for these, like before we would call getters on each field rather than merely reading from storage

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 2]
```diff
@@ -0,0 +1,1095 @@
...
+  }
+
+  /// Supply an asset only, and check no interest accrued.
+  function test_accrueInterest_OnlySupply(uint40 skipTime) public {
+    skipTime = uint40(bound(skipTime, 0, MAX_SKIP_TIME));
+    uint256 amount = 1000e18;
+    uint256 daiReserveId = _daiReserveId(spoke1);
+
+    // Bob supplies through spoke 1
+    Utils.supply(spoke1, daiReserveId, bob, amount, bob);
```
Comment (yan-man) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol:62]: hm seems skipTime isn't used? can add the skip here

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 3]
```diff
@@ -0,0 +1,1095 @@
...
+
+    _checkDebts(spoke1, daiReserveId, bob, 0, 0, 'after supply, no interest accrued');
+  }
+
+  function test_accrueInterest_BorrowAndWait() public {
```
Comment (yan-man) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: i think this test can just call test_accrueInterest_fuzz_BorrowAndWait with skipTime of 1 year

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 4]
```diff
@@ -0,0 +1,1095 @@
...
+    uint256 totalBase = _calcExpectedBaseDebt(amount, baseBorrowRate, startTime);
+    uint256 expectedPremiumDrawnShares = bobPosition.baseDrawnShares.percentMul(userRp);
+    uint256 expectedPremiumDebt = hub.convertToDrawnAssets(daiAssetId, expectedPremiumDrawnShares) -
+      bobPosition.premiumOffset +
+      bobPosition.realizedPremium;
+
+    _checkDebts(spoke1, daiReserveId, bob, totalBase, expectedPremiumDebt, 'after accrual');
+  }
+
+  function test_accrueInterest_fuzz_BorrowAndWait(uint40 skipTime) public {
```
Comment (yan-man) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: can just call test_accrueInterest_fuzz_BorrowAmountAndskipTime with inputs

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 5]
```diff
@@ -0,0 +1,1095 @@
...
+    uint256 totalBase = _calcExpectedBaseDebt(amount, baseBorrowRate, startTime);
+    uint256 expectedPremiumDrawnShares = bobPosition.baseDrawnShares.percentMul(userRp);
+    uint256 expectedPremiumDebt = hub.convertToDrawnAssets(daiAssetId, expectedPremiumDrawnShares) -
+      bobPosition.premiumOffset +
+      bobPosition.realizedPremium;
+
+    _checkDebts(spoke1, daiReserveId, bob, totalBase, expectedPremiumDebt, 'after accrual');
+  }
+
+  function test_accrueInterest_fuzz_BorrowAmountAndskipTime(
```
Comment (yan-man) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]:
```suggestion
  function test_accrueInterest_fuzz_BorrowAmountAndSkipTime(
```

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 6]
```diff
@@ -0,0 +1,1095 @@
...
+      expectedPremiumDrawnShares
+    ) -
+      bobPosition.premiumOffset +
+      bobPosition.realizedPremium;
+
+    _checkDebts(spoke1, usdxReserveId, bob, expectedBaseDebt, expectedPremiumDebt, 'after accrual');
+  }
+
+  // Fuzz a mix of borrowed and supplied assets for bob, check his RP, ensure correct interest accrual
+  function test_accrueInterest_fuzz_RPBorrowAndskipTime(
```
Comment (yan-man) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]:
```suggestion
  function test_accrueInterest_fuzz_RPBorrowAndSkipTime(
```

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 7]
```diff
@@ -0,0 +1,1095 @@
...
+import {Spoke} from 'src/contracts/Spoke.sol';
+import {LiquidityHub} from 'src/contracts/LiquidityHub.sol';
+
+contract SpokeAccrueInterestTest is SpokeBase {
+  using SharesMath for uint256;
+  using WadRayMath for uint256;
+  using WadRayMathExtended for uint256;
+  using PercentageMath for uint256;
+
+  uint256 public constant MAX_BPS = 1000_00;
```
Comment (yan-man) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: not sure if this is used

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 8]
```diff
@@ -0,0 +1,1095 @@
...
+      _wbtcReserveId(spoke1),
+      bob,
+      totalBase,
+      expectedPremiumDebt,
+      'wbtc after accrual'
+    );
+  }
+
+  // Fuzz a mix of borrowed and supplied assets for bob and rates, check his RP, ensure correct interest accrual
+  function test_accrueInterest_fuzz_RatesRPBorrowAndskipTime(
```
Comment (yan-man) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]:
```suggestion
  function test_accrueInterest_fuzz_RatesRPBorrowAndSkipTime(
```

File: tests/Base.t.sol (hunk 1 above)
Comment (CheyenneAtapour) [reply tests/Base.t.sol:873]: I was under the same impression @DhairyaSethi had thoughts about this. I'm fine to go either way

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 8 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Done

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 7 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Ty, removed

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 6 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Done

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 5 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Done

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 3 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: I just removed this altogether in favor of the other fuzz tests

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 4 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Just removed this altogether in favor of the other fuzz tests

File: tests/Base.t.sol (hunk 1 above)
Comment (yan-man) [reply tests/Base.t.sol:873]: oh did @DhairyaSethi disagree with this? thought he'd be in favor of the getters

File: tests/Base.t.sol [hunk 9]
```diff
@@ -982,6 +977,72 @@ abstract contract Base is Test {
     );
   }
 
+  function _checkDebts(
```
Comment (miguelmtzinf) [inline tests/Base.t.sol]: so this only covers the case where there is only 1 user and 1 spoke?
I d create 3 functions so they can be used across tests:
```
assertUserDebt(spoke, reserveId, user, expectedBaseDebt, expectedPremiumDebt);
assertReserveDebt(spoke, reserveId, expectedBaseDebt, expectedPremiumDebt);
assertAssetDebt(spoke, assetId, expectedBaseDebt, expectedPremiumDebt);
```

Comment (miguelmtzinf) [reply tests/Base.t.sol]: wondering if we should add a param for `expectedTotalDebt` or just assume it's the sum of both... but worth adding total debt check as well

File: tests/Base.t.sol [hunk 10]
```diff
@@ -1132,6 +1193,14 @@ abstract contract Base is Test {
     newBaseDebt = initialDrawnShares.rayMulUp(newDebtIndex);
   }
 
+  function _calcExpectedBaseDebt(
+    uint256 startingDebt,
+    uint256 borrowRate,
+    uint40 startTime
```
Comment (miguelmtzinf) [inline tests/Base.t.sol]:
```suggestion
  /// @dev Calculate expected base debt based on specified borrow rate
  function calculateExpectedBaseDebt(
    uint256 initialDebt,
    uint256 borrowRate,
    uint40 startTime
```
feels more aligned with the one above

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 11]
```diff
@@ -0,0 +1,1043 @@
...
+
+import 'tests/unit/Spoke/SpokeBase.t.sol';
+import {Spoke} from 'src/contracts/Spoke.sol';
+import {LiquidityHub} from 'src/contracts/LiquidityHub.sol';
+
+contract SpokeAccrueInterestTest is SpokeBase {
+  using SharesMath for uint256;
+  using WadRayMath for uint256;
+  using WadRayMathExtended for uint256;
+  using PercentageMath for uint256;
```
Comment (miguelmtzinf) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]:
```suggestion

contract SpokeAccrueInterestTest is SpokeBase {
  using SharesMath for uint256;
  using WadRayMathExtended for uint256;
  using PercentageMath for uint256;
```

Comment (miguelmtzinf) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: we gotta move to extended as much as we can

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 12]
```diff
@@ -0,0 +1,1043 @@
...
+    DataTypes.UserPosition memory bobPosition = spoke1.getUserPosition(usdxReserveId, bob);
+
+    uint256 expectedBaseDebt = _calcExpectedBaseDebt(borrowAmount, baseBorrowRate, startTime);
+    uint256 expectedPremiumDrawnShares = bobPosition.baseDrawnShares.percentMul(riskPremium);
+    uint256 expectedPremiumDebt = hub.convertToDrawnAssets(
+      usdxAssetId,
+      expectedPremiumDrawnShares
+    ) -
+      bobPosition.premiumOffset +
+      bobPosition.realizedPremium;
```
Comment (miguelmtzinf) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: i d be better if we compute this independently instead of using parts from local computation and onchain data.
smt like debtDelta*rp works

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 13]
```diff
@@ -0,0 +1,1043 @@
...
+      _deployLiquidity(
+        spoke1,
+        _wbtcReserveId(spoke1),
+        MAX_SUPPLY_AMOUNT - amounts.wbtcSupplyAmount
+      );
+    }
+
+    // Bob borrows dai from spoke 1
+    if (amounts.daiBorrowAmount > 0) {
+      Utils.borrow(spoke1, _daiReserveId(spoke1), bob, amounts.daiBorrowAmount, bob);
```
Comment (miguelmtzinf) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol:313]: wondering if we should move this logic inside the helper. We are doing this across all tests though

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 12 above)
Comment (miguelmtzinf) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: let's create a helper too pls

**Comment by miguelmtzinf** [comment]
can you please get these tests aligned with [LiquidityHubAccrueInterest.t.sol](https://github.com/aave/aave-v4/blob/main/tests/unit/LiquidityHub/LiquidityHubAccrueInterest.t.sol) ? Missing some cases, and also some names would be worth to keep the same.
Apart from that, we are checking debt side but not supply. I think we should add it if not too much effort

File: tests/Base.t.sol (hunk 9 above)
Comment (CheyenneAtapour) [reply tests/Base.t.sol]: I created 4 for all of them. I changed the original to reference a single user and am using that still (which calls all the new functions). Lemme know if that's an alright approach. Also added total debt check, assuming it will be the sum of both expected debts

File: tests/Base.t.sol (hunk 10 above)
Comment (CheyenneAtapour) [reply tests/Base.t.sol]: shouldn't we be using `_` to prefix these anyway, because these are internal functions? I applied the rest of the suggestions

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 11 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Done

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 12 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: I'm not so sure this works when there are 2 accruals, because the new premium is only applied to part of the debt. Also it introduces more precision loss than the current approach

**Comment by CheyenneAtapour** [comment]
> can you please get these tests aligned with [LiquidityHubAccrueInterest.t.sol](https://github.com/aave/aave-v4/blob/main/tests/unit/LiquidityHub/LiquidityHubAccrueInterest.t.sol) ? Missing some cases, and also some names would be worth to keep the same.

I saw 1 missing test case which I added. Some of the names are out of date because we use `skipTime` now instead of `elapsed`. Also some of the fuzz tests I felt were redundant, because other fuzz tests fuzz the same variables and more, so I removed the redundant ones. When we refactor `LiquidityHubAccrueInterest` tests, will align with these

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 13 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol:313]: I kinda liked keeping extra logic like this outside of utils, so we can also catch such edge cases in tests. I don't mind adding if you think it's better

File: tests/Base.t.sol (hunk 1 above)
Comment (DhairyaSethi) [reply tests/Base.t.sol:873]: yeah earlier the thinking was to reduce changes during code refactors which change the underlying accounting
now that's solidified, i was thinking we could have this getter return an adapted struct which also includes up to date debt and collateral balance values using the specific up to date getters

 it could clean up the tests slightly, but might be too much effort atm. wdyt?

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 14]
```diff
@@ -0,0 +1,1250 @@
...
+
+    // Check bob's base debt and premium debt for all assets at user, reserve, spoke, and asset level
+    DataTypes.UserPosition memory bobPosition = spoke1.getUserPosition(_daiReserveId(spoke1), bob);
+    totalBase = _calculateExpectedBaseDebt(
+      amounts.daiBorrowAmount,
+      rates.daiBaseBorrowRate,
+      startTime
+    );
+    uint256 expectedPremiumDrawnShares = bobPosition.baseDrawnShares.percentMul(bobRp);
+    uint256 expectedPremiumDebt = hub.convertToDrawnAssets(daiAssetId, expectedPremiumDrawnShares) -
```
Comment (DhairyaSethi) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: here we're still using the contract's debt ex rate as expected and asserting against the same, like expectedBaseDebt above we can reuse the index to get expectedPremiumDrawnAssets; but nbd

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 15]
```diff
@@ -0,0 +1,1250 @@
...
+          usingVirtualBalance: true
+        });
+      vm.mockCall(
+        address(irStrategy),
+        abi.encodeWithSelector(
+          IReserveInterestRateStrategy.calculateInterestRates.selector,
+          params
+        ),
+        abi.encode(rates.wethBaseBorrowRate)
+      );
```
Comment (DhairyaSethi) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: exact mocking feels overkill, can just partial match on calcInterestRate selector but nbd

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 16]
```diff
@@ -0,0 +1,1250 @@
...
+          hub.convertToDrawnAssets(daiAssetId, 1),
+          bob
+        );
+      }
+      // Construct mock call so we can see the same user rp calc as within the borrow function
+      vm.mockCall(
+        address(spoke2),
+        abi.encodeWithSelector(Spoke.getUserTotalDebt.selector, _daiReserveId(spoke2), bob),
+        abi.encode(spoke2.getUserTotalDebt(_daiReserveId(spoke2), bob) + 1e18) // Debt amount seen in the borrow function when calculating user rp
+      );
```
Comment (DhairyaSethi) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: this is the workaround for #421, right? would appreciate if we link issue on the comment in case we have a fix
also wondering why it works bc wouldn't this be needed for all other drawn positions of bob? bc of the notify?

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 17]
```diff
@@ -0,0 +1,1250 @@
...
+        bob,
+        amounts.wbtcBorrowAmount,
+        expectedPremiumDebt,
+        'wbtc before second accrual'
+      );
+      baseShares.wbtc = bobPosition.baseDrawnShares;
+
+      // Store index before accrual, and use this for calculating expected base debt
+      Indices memory indices;
+      indices.daiIndex = hub.getAsset(daiAssetId).baseDebtIndex;
```
Comment (DhairyaSethi) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: let's expose `hub.getNormalisedDebt(assetId)` like v3 so this is more future proof

Comment (DhairyaSethi) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: also all of these are not the up to date indexes here, that doesn't matter in the test below?

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 18]
```diff
@@ -0,0 +1,1250 @@
...
+      );
+
+      indices.usdxIndex = calculateExpectedDebtIndex(
+        indices.usdxIndex,
+        rates.usdxBaseBorrowRate,
+        beginningTime
+      );
+      bobPosition = spoke2.getUserPosition(_usdxReserveId(spoke2), bob);
+      totalBase = baseShares.usdx.rayMulUp(indices.usdxIndex);
+      expectedPremiumDrawnShares = baseShares.usdx.percentMul(bobRp);
```
Comment (DhairyaSethi) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: can we rename this to baseTotal or something? more straightforward and consistent with the existing nomenclature of the test having the type as suffix. but nbd

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 19]
```diff
@@ -0,0 +1,1250 @@
...
+
+      // Check bob's base debt and premium debt for all assets at user, reserve, spoke, and asset level
+      indices.daiIndex = calculateExpectedDebtIndex(
+        indices.daiIndex,
+        rates.daiBaseBorrowRate,
+        startTime
+      );
+      bobPosition = spoke2.getUserPosition(_daiReserveId(spoke2), bob);
+      totalBase = baseShares.dai.rayMulUp(indices.daiIndex);
+      expectedPremiumDrawnShares = baseShares.dai.percentMul(bobRp);
```
Comment (DhairyaSethi) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: do you think we should make this rounding direction explicit in this PR itself? would involve slight change on Spoke to set the exact rounding direction during premium calc (and potentially some other tests - if too many break we can avoid and tackle in a separate PR)

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 20]
```diff
@@ -0,0 +1,1250 @@
...
+    amounts.daiBorrowAmount = bound(amounts.daiBorrowAmount, 0, MAX_SUPPLY_AMOUNT / 2);
+    amounts.wethBorrowAmount = bound(amounts.wethBorrowAmount, 0, MAX_SUPPLY_AMOUNT / 2);
+    amounts.usdxBorrowAmount = bound(amounts.usdxBorrowAmount, 0, MAX_SUPPLY_AMOUNT / 2);
+    amounts.wbtcBorrowAmount = bound(amounts.wbtcBorrowAmount, 0, MAX_SUPPLY_AMOUNT / 2);
+
+    return amounts;
+  }
+
+  function _bound(Rates memory rates) internal pure returns (Rates memory) {
+    rates.daiBaseBorrowRate = bound(rates.daiBaseBorrowRate, 1, 100_00);
```
Comment (DhairyaSethi) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: can use constant from PercentageMath

Comment (DhairyaSethi) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: also max borrow rate can be higher than 100%, no?

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 21]
```diff
@@ -0,0 +1,1250 @@
...
+    rates.wethBaseBorrowRate = bound(rates.wethBaseBorrowRate, 1, 100_00);
+    rates.usdxBaseBorrowRate = bound(rates.usdxBaseBorrowRate, 1, 100_00);
+    rates.wbtcBaseBorrowRate = bound(rates.wbtcBaseBorrowRate, 1, 100_00);
+
+    // Put rates in ray
+    rates.daiBaseBorrowRate =
+      (rates.daiBaseBorrowRate * WadRayMathExtended.RAY) /
+      PercentageMath.PERCENTAGE_FACTOR;
```
Comment (DhairyaSethi) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: can use a test helper _bpsToRay for now, nbd

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 22]
```diff
@@ -0,0 +1,1250 @@
...
+        _getValueInBaseCurrency(wethAssetId, amounts.wethSupplyAmount) +
+        _getValueInBaseCurrency(usdxAssetId, amounts.usdxSupplyAmount) +
+        _getValueInBaseCurrency(wbtcAssetId, amounts.wbtcSupplyAmount) >
+        2 *
+          (_getValueInBaseCurrency(daiAssetId, amounts.daiBorrowAmount) +
+            _getValueInBaseCurrency(wethAssetId, amounts.wethBorrowAmount) +
+            _getValueInBaseCurrency(usdxAssetId, amounts.usdxBorrowAmount) +
+            _getValueInBaseCurrency(wbtcAssetId, amounts.wbtcBorrowAmount))
+    );
+  }
```
Comment (DhairyaSethi) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: ah man i hope this doesn't make the test reject on higher fuzz runs 🫠

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 12 above)
Comment (DhairyaSethi) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: we can use cached indexes maybe?

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 16 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Users only have 1 risk premium, and it's calculated using this mocked value, so the rest use this same calculated risk premium

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 22 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: We could easily handle by just returning out of the test in this case if we run into that issue. E.g. Switch this function to return a boolean and then return out of the function accordingly if this condition isn't met

Comment (DhairyaSethi) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: return exits the test, wastes a fuzz run and give a false sense of covered cases bc it's counted as a passing test. preference would be bound params instead

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 16 above)
Comment (DhairyaSethi) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: yes but my point is on the debt positions. there will be mismatch will be for all others regardless, no?

Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: All the debt positions are computed using this "mismatched" user RP, which we are accounting for here by calculating it how the protocol sees it when it calculates

Comment (DhairyaSethi) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: right so the other way around, makes sense ty! 
it's unfortunate we were seeing very high differences in premium debt bc of the token decimal otherwise approxEq could have worked here. 
sounds good 

(nit: pls encodeCall over encodeWithSelector)

File: tests/Base.t.sol (hunk 1 above)
Comment (CheyenneAtapour) [reply tests/Base.t.sol:873]: Maybe we can do that on a separate PR related to test cleanup

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 16 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Done

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 17 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Aren't they all up to date? Because an action just happened on the same block as using these getters

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 18 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: I'm not a fan of baseTotal, but it could also just be `baseDebt`

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 20 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Yeah, I now used the max from irStrategy

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 21 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Done

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 20 above)
Comment (DhairyaSethi) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: we should note to do that for all existing tests

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 12 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: @miguelmtzinf Indeed, I see a precision delta of 12 wei with that approach, even with 1 accrual, but I've added this approach with the helper to all but 2 tests

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 16 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: I added the comment about it being a workaround, sorry missed that earlier

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 20 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Cool, I just did a quick look and didn't immediately see other tests that weren't doing this. So I made an issue: https://github.com/orgs/aave/projects/14?pane=issue&itemId=114073962&issue=aave%7Caave-v4%7C437

Currently, we define again the constant MAX_BORROW_RATE in our tests to match the value in the `irStrategy`. I mention in the above linked issue that we can instead pull directly from that contract, so it's more robust / future proof

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 17 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Ah, wait, you mean after the time skip. Indeed I calculate what the expected debt index should be, which is part of what I'm testing

Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: On second thought, I'm fine to expose that function, but there's something nice about what I'm testing. Since the assets besides dai haven't been interacted with, I'm calculating their expected indices from the _beginning_ time t0 of this test. There are comments about this in the following calculations

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 19 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: This seems to cause some issues and even underflow, if I'm understanding correctly. Also feels best for another PR

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 22 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: Alright, changed this to a bound

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 13 above)
Comment (miguelmtzinf) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol:313]: which edge cases are you thinking about?

Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol:313]: For example, reverting on trying to borrow 0. However, I suppose we can and should test such edge cases in the specific test suite where we don't use the Utils and instead directly call the function

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 19 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol]: This is now being done in https://github.com/aave/aave-v4/pull/440 and hopefully can improve the assertions on this PR

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 23]
```diff
@@ -0,0 +1,1110 @@
...
+    );
+
+    bobPosition = spoke1.getUserPosition(_usdxReserveId(spoke1), bob);
+    baseDebt = _calculateExpectedBaseDebt(
+      amounts.usdxBorrowAmount,
+      rates.usdxBaseBorrowRate,
+      startTime
+    );
+    expectedPremiumDrawnShares = bobPosition.baseDrawnShares.percentMulUp(bobRp);
+    expectedPremiumDebt =
```
Comment (miguelmtzinf) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol:940]: we still cannot use calc helpers?

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol [hunk 24]
```diff
@@ -0,0 +1,1110 @@
...
+    );
+
+    bobPosition = spoke1.getUserPosition(_wbtcReserveId(spoke1), bob);
+    baseDebt = _calculateExpectedBaseDebt(
+      amounts.wbtcBorrowAmount,
+      rates.wbtcBaseBorrowRate,
+      startTime
+    );
+    expectedPremiumDrawnShares = bobPosition.baseDrawnShares.percentMulUp(bobRp);
+    expectedPremiumDebt =
```
Comment (miguelmtzinf) [inline tests/unit/Spoke/Spoke.AccrueInterest.t.sol:986]: here too

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 23 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol:940]: Created an issue here: https://github.com/orgs/aave/projects/14/views/1?pane=issue&itemId=116146155&issue=aave%7Caave-v4%7C466

File: tests/unit/Spoke/Spoke.AccrueInterest.t.sol (hunk 24 above)
Comment (CheyenneAtapour) [reply tests/unit/Spoke/Spoke.AccrueInterest.t.sol:986]: Same issue as above
